chef_master/source/deprecations_chef_platform_methods.rst

```diff
@@ -0,0 +1,112 @@
+=============================================
+Deprecation: Chef::Platform methods (CHEF-13)
+=============================================
+`[edit on GitHub] <https://github.com/chef/chef-web-docs/blob/master/chef_master/source/deprecations_chef_platform_methods.rst>`__
+
+.. tag deprecations_chef_platform_methods
+
+Several methods under `Chef::Platform` that were previously public APIs to control resolution of provider classes were replaced by the dyanmic
+`Chef::ProviderResolver` work and the `provides` keyword.
+.. end_tag
+
+This deprecation warning was added in Chef 12.18.x, and using these APIs will become a hard error in Chef 13.
+
+Remediation
+================
+
+Code which used to use `Chef::Platform.provider_for_resource` or `Chef::Platform.find_provider` to create providers for a resource:
+
+.. code-block:: ruby
+
+  resource = Chef::Resource::File.new("/tmp/foo.xyz", run_context)
+  provider = Chef::Platform.provider_for_resource(resource, :create)
+
+  resource = Chef::Resource::File.new("/tmp/foo.xyz", run_context)
+  provider = Chef::Platform.find_provider("ubuntu", "16.04", resource)
+
+  resource = Chef::Resource::File.new("/tmp/foo.xyz", run_context)
+  provider = Chef::Platform.find_provider_for_node(node, resource)
+
+Should instead use the `Chef::Resource#provider_for_action` API on the instance of the resource:
+
+.. code-block:: ruby
+
+  resource = Chef::Resource::File.new("/tmp/foo.xyz", run_context)
+  provider = resource.provider_for_action(:create)
+
+As the internal resources and providers in core chef have been ported over to use the `Chef::ProviderResolver` dynamic resolution the use
+of the old Chef::Platform class methods have actually been broken.  Tools like `chefspec` and `chef-minitest-handler` were ported over to
+the new APIs in Chef 12.0.  The `Chef::Resource#provider_for_action` API dates back to before Chef 11.0.0 and is fully backwards compatible,
+any remaining code using the old APIs should be exceedingly buggy at this point.
+
+Also, code which used to use `Chef::Platform.set` to register providers for a platform/platform_version should use the `provides` keyword
+on the provider instead:
+
+.. code-block:: ruby
+  Chef::Platform.set platform: :fedora, version: '>= 19', resource: :mysql_service, provider: Chef::Provider::MysqlServiceSystemd
+
+Should be replaced by:
+
+.. code-block:: ruby
+  class Chef::Provider::MysqlSserviceSystemd
```

Comment:
Need a return between the `code-block` directive and the first line of the code itself.